persistence/sql/persister_oauth2.go

```diff
@@ -692,18 +692,20 @@ func (p *Persister) InvalidateUserCodeSession(ctx context.Context, signature str
 	)
 }
 
-// UpdateAndInvalidateUserCodeSession invalidates a user code session and connects it with the device flow challenge ID
-func (p *Persister) UpdateAndInvalidateUserCodeSession(ctx context.Context, signature, challenge_id string) (err error) {
+// UpdateAndInvalidateUserCodeSession invalidates a user code session and connects it with the device flow request ID
+func (p *Persister) UpdateAndInvalidateUserCodeSessionByRequestID(ctx context.Context, request_id, challenge_id string) (err error) {
 	ctx, span := p.r.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.UpdateAndInvalidateUserCodeSession")
 	defer otelx.End(span, &err)
 
+	// TODO(nsklikas): afaict this is supposed to return an error if no rows were updated, but this is not the actual behavior.
+	// We need to either fix this OR do a select -> check -> update (this would require 2 queries instead of 1).
```

Comment:
would we be able to use the `insert on conflict update set` instead?would that fit our case?

Reply:
The problem is that if the update returns 0 rows, that means that `active=False`. Which means that the user code has already been used. In that case we would like to rollback the transaction, so that the flow is not written to the database (before we invalidate the user code, we persist the flow to the database on https://github.com/canonical/hydra/pull/18/files#diff-265e72885dd37a9d8c6f81e2c928b2bfad035cecd98051ef2f04ea88a56ed540R1237).

With the current code, even if we add a transaction we won't be able to rollback because we won't get an error. Maybe I am missing something, because `sqlcon.HandleError` is supposed to handle the case of 0 rows being returned (but afaict gobuffalo does not return an error).